pkg/bitlpm/trie.go

```diff
@@ -0,0 +1,462 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package bitlpm
+
+// Trie is a non-thread-safe binary trie that indexes arbitrarily long
+// bit-based keys with associated prefixes (that is, "masks") indexed from
+// [most significant bit] ("MSB") to [least significant bit] ("LSB") using
+// the [longest prefix match algorithm].
+//
+// Each method's comments describes the mechanism of how the method
+// works.
+//
+// [most significant bit]: https://en.wikipedia.org/wiki/Bit_numbering#Most_significant_bit
+// [least significant bit]: https://en.wikipedia.org/wiki/Bit_numbering#Least_significant_bit
+// [longest prefix match algorithm]: https://en.wikipedia.org/wiki/Longest_prefix_match
+type Trie[K, T any] interface {
```

Comment:
Hmm, why is this an interface? Will we ever need to dynamically swap out implementations? I'd rather it just be methods on a struct if possible.

Reply:
Chicken-and-egg when dealing with allowing both a Trie that accepts user-defined `Key[K]` arguments and library-defined keys like `Unsigned` while keeping the method signature the same.